### PR TITLE
samples: Bluetooth: Add BLE Periodic Advertising sample

### PR DESCRIPTION
**Pull request by cvinayak**
Add an application that initializes Bluetooth Subsystem,
creates an extended advertising set, and starts Periodic
Advertising functionality.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

### CODE REVIEW

File: include/bluetooth/bluetooth.h
Comment (Thalley) [inline include/bluetooth/bluetooth.h:656]: I was actually considering to add these as well, as I was missing them as well :D Glad to see them added

File: samples/bluetooth/per_adv/src/main.c
Comment (Thalley) [inline samples/bluetooth/per_adv/src/main.c:14]: Seeing as you added the other macros for extended adv, perhaps similar macros for the ext_adv_start_param could be added? I suppose most adv will just advertise without num_events or timeout so it would be generally useful. Can be added in another PR.

**Review by Thalley — APPROVED** [review]
LGTM

File: samples/bluetooth/per_adv/src/main.c
Comment (cvinayak) [reply samples/bluetooth/per_adv/src/main.c:14]: I thought of adding, but as the Zephyr controller does not support connectable over extended advertising yet, I dont have/know any means to test it manually or in CI, hence not included in this PR. Yes, can be added in another PR.

Comment (Thalley) [reply samples/bluetooth/per_adv/src/main.c:14]: I'm sure I understand: What does the `bt_le_ext_adv_start_param` have to do with connectable advertising? I guess you could test such a macro with the sample added in this PR, couldn't you?

Comment (cvinayak) [reply samples/bluetooth/per_adv/src/main.c:14]: Correction, my mistake, I was thinking of `BT_LE_ADV_CONN`et al. 

Yes, `bt_le_ext_adv_start_param` can have a define. Will give a try.

**Review by Thalley — APPROVED** [review]
LGTM

File: include/bluetooth/bluetooth.h
Comment (rugeGerritsen) [inline include/bluetooth/bluetooth.h:644]: Now that you are already adding these new APIs, maybe you should also add `BT_LE_EXT_ADV_NCONN_CODED`?
Maybe also apply the same pattern to `BT_LE_EXT_ADV_NCONN_NAME` and `BT_LE_EXT_ADV_NCONN_IDENTITY`

**Review by rugeGerritsen — COMMENTED** [review]
Maybe we should split this review into two parts? The ext. adv. API updates are not strictly related to the sample itself. 

Should the sample documentation include information on how to test it? Now the only user feedback is "Starting Periodic Advertising Demo". Are there phones or other devices that can intercept these periodic advertising packets?
I would assume the sample documentation needs some more information on how it can be used or validated by the user.

**Comment by cvinayak** [comment]
> Maybe we should split this review into two parts? The ext. adv. API updates are not strictly related to the sample itself.

Keeping the defines and samples in separate PR does not make sense, as PR just for defines not used anywhere would need a sample anyway for testing.

> Should the sample documentation include information on how to test it? Now the only user feedback is "Starting Periodic Advertising Demo". Are there phones or other devices that can intercept these periodic advertising packets?
I would assume the sample documentation needs some more information on how it can be used or validated by the user.

Yes, will update the documentation as I send PRs for sync sample. This is still a draft PR and commits to add a sample to generate periodic advertising report will make it big (still waiting for other open PRs to be merged). Keeping all the code in single PR will make it more understandable, but it is not desired to have big PRs.

File: include/bluetooth/bluetooth.h
Comment (cvinayak) [reply include/bluetooth/bluetooth.h:644]: I now have included the sample to sync with the periodic advertising, will add the defines for coded PHY and push the changes once I am done testing.

**Comment by cvinayak** [comment]
> the sample documentation needs some more information on how it can be used or validated by the user.

Yes, will include more information and update the PR soon.

File: samples/bluetooth/per_sync/src/main.c
Comment (Thalley) [inline samples/bluetooth/per_sync/src/main.c:268]: Since this is within a do { ... } while, could you not remove the goto and the label and use `continue` instead?

File: samples/bluetooth/per_sync/src/main.c
Comment (Thalley) [inline samples/bluetooth/per_sync/src/main.c:261]: Did you consider putting this part in the scan callback? So rather than using a semaphore and a global variable, you just create the sync in the callback instead.

Comment (cvinayak) [reply samples/bluetooth/per_sync/src/main.c:260]: Calling APIs in callback increases internal rx thread call stack usage. Keeping the API calls together improves readability in terms of demonstrating API usage in this sample.

File: samples/bluetooth/per_sync/src/main.c
Comment (cvinayak) [reply samples/bluetooth/per_sync/src/main.c]: Replaced code to use `continue`

File: samples/bluetooth/per_sync/src/main.c
Comment (rugeGerritsen) [inline samples/bluetooth/per_sync/src/main.c:84]: This print takes two full lines, when displaying this print on my terminal with the default terminal size configuration.
That makes it harder to to see what is happing. 

For me it would have been sufficient to only print address and device name.

File: samples/bluetooth/per_sync/src/main.c
Comment (rugeGerritsen) [inline samples/bluetooth/per_sync/src/main.c:248]: Did you consider stopping the scanner after the sync was established? 
I find it hard to see the prints related to the periodic advertiser in terminal output. It is overflowed by advertising reports from the extended advertiser.

File: samples/bluetooth/per_sync/src/main.c
Comment (rugeGerritsen) [inline samples/bluetooth/per_sync/src/main.c:225]: Strictly speaking you don't need to perform active scanning when scanning for extended advertisers. I'm fine with keeping it as is.

File: samples/bluetooth/per_sync/src/main.c
Comment (rugeGerritsen) [inline samples/bluetooth/per_sync/src/main.c:146]: To me the interesting parts is seeing that that the data is being updated. Therefore I would prioritize printing the data instead of printing the `tx_power` and `cte` field. Those are anyways set to "not present" for this sample

File: samples/bluetooth/per_sync/src/main.c
Comment (rugeGerritsen) [inline samples/bluetooth/per_sync/src/main.c:266]: I assume that we in all cases should return from main when we fail to take a semaphore with infinite timeout?
```suggestion
			return;
```

File: samples/bluetooth/per_sync/src/main.c
Comment (rugeGerritsen) [inline samples/bluetooth/per_sync/src/main.c:201]: To simplify the sample, you could consider using static initialization of the semaphores using `K_SEM_DEFINE()`

**Review by rugeGerritsen — COMMENTED** [review]
I would prefer the samples to be named `periodic_adv` and `periodic_sync` to be more explicit. Users with only basic understanding of Bluetooth LE may not know if `per` is an abbreviation for `peripheral` or `periodic`.

File: include/bluetooth/hci.h
Comment (cvinayak) [reply include/bluetooth/hci.h:1970]: `RSSI is not available` with value 0x7F is described only for 4 Bluetooth LE events. All other events use RSSI value range. Correct me, if I have missed non-LE events that use the 0x7F explicit value, HCI_Read_RSSI uses 0x81 to indicate signal strength indication is not available.

File: samples/bluetooth/per_adv/boards/nrf52840dk_nrf52840.conf
Comment (cvinayak) [reply samples/bluetooth/per_adv/boards/nrf52840dk_nrf52840.conf:1]: Not all boards use the native split controller.

File: samples/bluetooth/per_adv/src/main.c
Comment (cvinayak) [reply samples/bluetooth/per_adv/src/main.c:9]: Manufacturer data, I just used this from the `scan_adv` sample.

File: samples/bluetooth/per_adv/src/main.c
Comment (cvinayak) [reply samples/bluetooth/per_adv/src/main.c:12]: The `const` applies to the `ad` memory, `mfg_data` memory is non-const.

File: samples/bluetooth/per_sync/src/main.c
Comment (cvinayak) [reply samples/bluetooth/per_sync/src/main.c:28]: `is_per_adv`: There `is` a `per`iodic `adv`ertising.

File: samples/bluetooth/per_sync/src/main.c
Comment (cvinayak) [reply samples/bluetooth/per_sync/src/main.c:84]: This is taken from the shell application.

File: samples/bluetooth/per_sync/src/main.c
Comment (cvinayak) [reply samples/bluetooth/per_sync/src/main.c:248]: No, so as to be able to demo that periodic sync continues even when local events could be skipped due to scan events or possible extended advertising collisions on auxiliary channels from other devices.

But yes, if this is a sample to demo low power usage of periodic adv sync, scanning be disabled after sync established.

File: samples/bluetooth/per_sync/src/main.c
Comment (cvinayak) [reply samples/bluetooth/per_sync/src/main.c:146]: As this sample establishes periodic adv sync to any periodic advertising in the vicinity, I will leave it to the user to parse the data as he desires, but yes, I will add display of the manufacturer data if present.

action: display of the manufacturer data.

File: samples/bluetooth/per_sync/src/main.c
Comment (cvinayak) [reply samples/bluetooth/per_sync/src/main.c:266]: action: I will use a timeout and cancel sync create. It will be difficult to test (to kill periodic advertiser before the scanner sync after it has found it).

File: include/bluetooth/hci.h
Comment (rugeGerritsen) [reply include/bluetooth/hci.h:1970]: HCI_Read_RSSI: "If the RSSI cannot be read the RSSI metric shall be set to 127 (0x7F)"

File: samples/bluetooth/per_adv/src/main.c
Comment (rugeGerritsen) [reply samples/bluetooth/per_adv/src/main.c:9]: What is the `g` abbreviating? Maybe naming it something as simple as `ad_data` is better?

File: samples/bluetooth/per_sync/src/main.c
Comment (Thalley) [reply samples/bluetooth/per_sync/src/main.c:28]: I guess it could be renamed into `per_adv_found` if that helps :)

Comment (cvinayak) [reply samples/bluetooth/per_sync/src/main.c]: done

File: samples/bluetooth/per_sync/src/main.c
Comment (cvinayak) [reply samples/bluetooth/per_sync/src/main.c]: done

File: samples/bluetooth/per_sync/src/main.c
Comment (cvinayak) [reply samples/bluetooth/per_sync/src/main.c]: done

File: samples/bluetooth/per_sync/src/main.c
Comment (cvinayak) [reply samples/bluetooth/per_sync/src/main.c]: No action, as the sample will be able to demonstrate active extended scanning co-existing with periodic advertising synchronization. (yes, periodic advertising sync does not need active scanning)

File: samples/bluetooth/per_sync/src/main.c
Comment (cvinayak) [reply samples/bluetooth/per_sync/src/main.c]: done

File: include/bluetooth/hci.h
Comment (cvinayak) [reply include/bluetooth/hci.h]: Vol 5: AMP Host Controller Interface, Section 2.6, `The value shall be 0x81 (-127 dBm) if the signal strength indication is not
available.` Hence, retaining `BT_HCI_LE_RSSI_NOT_AVAILABLE` to specify LE HCI event value.

**Comment by carlescufi** [comment]
@joerchan @jhedberg @rugeGerritsen could you review?

**Review by jhedberg — APPROVED** [review]

**Comment by Moediddy901** [comment]
I need help

**Pull request merged**
